packaging/metadata.py

```diff
@@ -0,0 +1,191 @@
+    KEYWORDS = "keywords"
+    HOME_PAGE = "home-page"
+    AUTHOR = "author"
+    AUTHOR_EMAIL = "author-email"
+    LICENSE = "license"
+    # 1.1
+    SUPPORTED_PLATFORM = "supported-platform"
+    DOWNLOAD_URL = "download-url"
+    CLASSIFIER = "classifier"
+    # 1.2
+    MAINTAINER = "maintainer"
+    MAINTAINER_EMAIL = "maintainer-email"
+    REQUIRES_DIST = "requires-dist"
+    REQUIRES_PYTHON = "requires-python"
+    REQUIRES_EXTERNAL = "requires-external"
+    PROJECT_URL = "project-url"
+    PROVIDES_DIST = "provides-dist"
+    OBSOLETES_DIST = "obsoletes-dist"
+    # 2.1
+    DESCRIPTION_CONTENT_TYPE = "description-content-type"
+    PROVIDES_EXTRA = "provides-extra"
+
+
+class Metadata:
+
+    """
+    A representation of core metadata.
+    """
+
+    # A property named `display_name` exposes the value.
+    _display_name: str
+    # A property named `canonical_name` exposes the value.
+    _canonical_name: utils.NormalizedName
+    version: packaging_version.Version
+    platforms: Set[str]
+    summary: str
+    description: str
+    keywords: List[str]
+    home_page: str
+    author: str
+    author_emails: List[_NameAndEmail]
+    license: str
+    supported_platforms: Set[str]
+    download_url: str
+    classifiers: Set[str]
+    maintainer: str
+    maintainer_emails: List[_NameAndEmail]
+    requires_dists: Set[requirements.Requirement]
+    requires_python: specifiers.SpecifierSet
+    requires_externals: Set[str]
+    project_urls: Set[_LabelAndURL]
+    provides_dists: Set[requirements.Requirement]
+    obsoletes_dists: Set[requirements.Requirement]
+    description_content_type: str
+    provides_extras: Set[utils.NormalizedName]
+    dynamic: Set[DynamicField]
+
+    def __init__(
+        self,
+        name: str,
+        version: packaging_version.Version,
+        *,
+        # 1.0
+        platforms: Optional[Iterable[str]] = None,
+        summary: Optional[str] = None,
+        description: Optional[str] = None,
+        keywords: Optional[Iterable[str]] = None,
+        home_page: Optional[str] = None,
+        author: Optional[str] = None,
+        author_emails: Optional[Iterable[_NameAndEmail]] = None,
+        license: Optional[str] = None,
+        # 1.1
+        supported_platforms: Optional[Iterable[str]] = None,
+        download_url: Optional[str] = None,
+        classifiers: Optional[Iterable[str]] = None,  # TODO: OK?
+        # 1.2
+        maintainer: Optional[str] = None,
+        maintainer_emails: Optional[Iterable[_NameAndEmail]] = None,
+        requires_dists: Optional[Iterable[requirements.Requirement]] = None,
+        requires_python: Optional[specifiers.SpecifierSet] = None,
+        requires_externals: Optional[Iterable[str]] = None,  # TODO: OK?
+        project_urls: Optional[Iterable[_LabelAndURL]] = None,
+        provides_dists: Optional[Iterable[requirements.Requirement]] = None,
+        obsoletes_dists: Optional[Iterable[requirements.Requirement]] = None,
```

Comment:
Nobody uses obsoletes_dist, and there's only *one* case of provides_dist, which is "frobozz (0.1)" (not a valid Requirement). I'd be tempted to leave these as `str` rather than `Requirement`.

Reply:
Are you thinking that because of the overhead, or you just want to maximize compatibility by simply not dealing with the field as much as possible?

Reply:
1. My biggest use case is manipulating data that currently exists on PyPI, and so unless it causes other issues, I prefer to be forgiving in what we accept. Unlike (for example) versions, where being strict has significant benefits (because there's semantics involved) nobody actually uses `(provides|obsoletes)_dist`, and they are defined in the spec as "rarely used", so I don't think there's a lot of value in requiring the stricter `Requirement`.
2. Literally the only example of either of these fields in *every* wheel on PyPI has a value of `frobozz (0.1)` which is not a valid requirement! Is it really sensible to mandate a form which the only real-world example doesn't follow? 🙂 I concede that the example is probably intended to be `frobozz (==0.1)` and may even be valid legacy form, but `Requirement` can't parse it...
3. I'm not actually sure the semantics of making it a requirement make sense anyway - what does it mean to say foo provides "bar != 2.0; os_name=='nt'"? The spec allows this (and even has a similar example) but honestly, does it make any sense? I'd prefer to use `str` for now, if only to protect against any future change/clarification in the semantics of the field.

Honestly, though, it's two essentially unused fields, I'm not going to stress either way (but see my comment below for further on this).